exchanges/bitfinex/bitfinex_websocket.go

```diff
@@ -869,139 +873,99 @@ func (b *Bitfinex) handleWSTickerUpdate(c *subscription.Subscription, d []interf
-		if !ok {
-			return errors.New("unable to type assert trade snapshot data")
-		}
-		for i := range snapshot {
-			elem, ok := snapshot[i].([]interface{})
-			if !ok {
-				return errors.New("unable to type assert trade snapshot element data")
-			}
-			tradeID, ok := elem[0].(float64)
-			if !ok {
-				return errors.New("unable to type assert trade ID")
-			}
-			timestamp, ok := elem[1].(float64)
-			if !ok {
-				return errors.New("unable to type assert trade timestamp")
-			}
-			amount, ok := elem[2].(float64)
-			if !ok {
-				return errors.New("unable to type assert trade amount")
-			}
-			wsTrade := WebsocketTrade{
-				ID:        int64(tradeID),
-				Timestamp: int64(timestamp),
-				Amount:    amount,
-			}
-			if len(elem) == 5 {
-				rate, ok := elem[3].(float64)
-				if !ok {
-					return errors.New("unable to type assert trade rate")
-				}
-				wsTrade.Rate = rate
-				period, ok := elem[4].(float64)
-				if !ok {
-					return errors.New("unable to type assert trade period")
-				}
-				wsTrade.Period = int64(period)
-			} else {
-				price, ok := elem[3].(float64)
-				if !ok {
-					return errors.New("unable to type assert trade price")
-				}
-				wsTrade.Rate = price
-			}
-			tradeHolder = append(tradeHolder, wsTrade)
-		}
-	case 3:
-		if eventType != wsFundingTradeUpdate && eventType != wsTradeExecutionUpdate {
-			return fmt.Errorf("unhandled WS trade update event: %s", eventType)
+	var wsTrades []*wsTrade
+	switch valueType {
+	case jsonparser.String:
+		if t, err := b.handleWSPublicTradeUpdate(respRaw); err != nil {
+			return err
+		} else {
+			wsTrades = []*wsTrade{t}
 		}
-		data, ok := d[2].([]interface{})
-		if !ok {
-			return errors.New("trade data type assertion error")
+	case jsonparser.Array:
+		if wsTrades, err = b.handleWSPublicTradesSnapshot(respRaw); err != nil {
+			return err
 		}
-
-		tradeID, ok := data[0].(float64)
-		if !ok {
-			return errors.New("unable to type assert trade ID")
+	default:
+		return fmt.Errorf("%w `tradesUpdate[1]`: %w `%s`", errParsingWSField, jsonparser.UnknownValueTypeError, valueType)
+	}
+	trades := make([]trade.Data, len(wsTrades))
+	for i, w := range wsTrades {
+		t := trade.Data{
+			Exchange:     b.Name,
+			AssetType:    s.Asset,
+			CurrencyPair: s.Pairs[0],
+			TID:          strconv.FormatInt(w.ID, 10),
+			Timestamp:    w.Timestamp.Time().UTC(),
+			Side:         order.Buy,
+			Amount:       w.Amount,
+			Price:        w.Price,
 		}
-		timestamp, ok := data[1].(float64)
-		if !ok {
-			return errors.New("unable to type assert trade timestamp")
+		if w.Period != 0 {
+			t.AssetType = asset.MarginFunding
 		}
-		amount, ok := data[2].(float64)
-		if !ok {
-			return errors.New("unable to type assert trade amount")
+		if t.Amount < 0 {
+			t.Side = order.Sell
+			t.Amount = math.Abs(t.Amount)
 		}
-		wsTrade := WebsocketTrade{
-			ID:        int64(tradeID),
-			Timestamp: int64(timestamp),
-			Amount:    amount,
+		if feedEnabled {
+			b.Websocket.DataHandler <- &t
```

Comment:
With verbose on; also we should send it over as a slice instead of individually. 
```
[WARN]  | WEBSOCKET | 29/12/2024 09:21:31 | Bitfinex websocket Unknown type: &{ID:00000000-0000-0000-0000-000000000000 TID:1695543964 Exchange:Bitfinex CurrencyPair:ETHUSD AssetType:spot Side:SELL Price:3414.9 Amount:1.01092617 Timestamp:2024-12-28 22:21:31.398 +0000 UTC}
```

Reply:
This was  a very deliberate change from me.
Consumers of trade updates want to get trades, not slices of varying lengths of trades.
It makes a consumer's life unnecessarily harder to say: _Oh, I read something from a channel. Is it a thing, or a slice of those things?_
Unless the slice is integral, like a related tuple or an orderbook, I definitely believe we should send things, not slices of things.

I could go change engine to accept a pointer to a trade instead of a slice.
But at the same time, I don't want to introduce inconsistency across the exchanges.
So I'll revert my change, and then switch over to individual trades in a separate PR later.

Fixed https://github.com/gbjk/gocryptotrader/commit/48ebaf7be

Reply:
Right that does make consumer logic more simple. I was more coming from the point of view reducing the channel contention which adds a little bit of latency between handovers. Your way is more maintainable maybe in future we can have a configurable layer which you can process both but for such an edge case, it's doubtful.